### ISSUE TITLE
No errors for keyword arguments in class statement

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#6403]
Example:

```python
from typing import TypedDict

class TD(TypedDict, a=3):
    a: int
```

https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFC1UA2AhgM6uHkAURCZlNADRRmAXgDMASgBctKPJHTMKGEA

This fails at runtime, and type checkers should reject it too. (cf. python/mypy#16438)

**Comment by erictraut** [comment]
This is currently by design. Pyright doesn't ever report extra keyword arguments passed in a class declaration. So this issue isn't specific to `TypedDict`.

I'm going to change this from a bug to an enhancement request and consider it in that light. I think it's a reasonable feature to add, assuming it doesn't produce too many false positives in typical code bases.

**Comment by JelleZijlstra** [comment]
Thanks, I didn't know that. For normal classes, keyword arguments may be valid, depending on the metaclass and `__init_subclass__` in base classes. It would be good if pyright could check keyword arguments for correctness, but it's not simple. For TypedDict classes, however, any keyword arguments are always invalid.

**Comment by erictraut** [comment]
Currently, pyright does check `__init_subclass__` if it's defined somewhere in the class hierarchy _other than the `object` class_. The `object.__init_subclass__` method is ignored. I don't recall why I did it this way. It may have been because of `TypedDict`, which is a special form and doesn't define an `__init_subclass__` or a metaclass.

**Comment by erictraut** [comment]
The reason pyright is not flagging the error for a `TypedDict` is because of a faulty type definition in `typing.pyi` and `typing_extensions.pyi`. I suspect the same is true for mypy. Both of these stubs define `_TypedDict` with a metaclass `ABCMeta` (which is a lie — the actual runtime metaclass is `_TypedDictMeta`). `ABCMeta`'s `__new__` method accepts any keyword arguments, which is why you're not seeing an error in your code sample above.

I propose that we fix this by modifying the definition of `_TypedDict` as follows.

```python
@type_check_only
class _TypedDictMeta(type):
    def __new__(
        __mcls: type[typing_extensions.Self],
        __name: str,
        __bases: tuple[type, ...],
        __namespace: dict[str, Any],
        *,
        total: bool = False,
    ) -> typing_extensions.Self: ...

@type_check_only
class _TypedDict(Mapping[str, object], metaclass=_TypedDictMeta):
    ...
```

This not only addresses the bug and more closely reflects the actual runtime class hierarchy, but it also has the benefit of providing better in-editor support for the `total` parameter.

Thoughts?

**Comment by JelleZijlstra** [comment]
Good catch, that sounds like a good idea!

Type checkers may still need a special case to prohibit a custom `metaclass=` on TypedDict, but your suggestion should automatically prohibit other keyword arguments.

**Comment by erictraut** [comment]
I've improved pyright's handling of `__init_subclass__` to handle the case where there is no custom metaclass with a `__new__` method and no `__init_subclass__` that overrides `object.__init_subclass__`. I've also improved the error messages for `__init_subclass__` in general. This set of changes still doesn't result in an error for the code in the original sample above, but that will be fixed once the `_TypedDict` definition is updated in typeshed.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.336, which I just published. It will also be included in a future release of pylance.
